### PR TITLE
Add new column to spanner schema

### PR DESCRIPTION
**Pull request by shreyakhajanchi**
1. Add new column form
2. Skip new column during data migration
3. Show the added column in the spanner draft along with edit capabilities
4. Give option to restore dropped columns that exist in source not new columns

### CODE REVIEW

File: internal/helpers.go [hunk 1]
```diff
@@ -140,12 +141,12 @@ func GetSrcIndexFromId(indexes []schema.Index, indexId string) (schema.Index, er
 func ComputeUsedNames(conv *Conv) map[string]bool {
 	usedNames := make(map[string]bool)
 	for _, table := range conv.SpSchema {
-		usedNames[table.Name] = true
+		usedNames[strings.ToLower(table.Name)] = true
```
Comment (bharadwaj-aditya) [inline internal/helpers.go:144]: is this being done throughout the code ? How was it working before this PR as the older code snippet does not move to lower.

File: sources/common/infoschema.go [hunk 2]
```diff
@@ -120,8 +120,8 @@ func ProcessData(conv *internal.Conv, infoSchema InfoSchema) {
 				srcSchema.Name, ok))
 			continue
 		}
-		// Extract spColds without synthetic primary key columnn id.
-		colIds := RemoveSynthId(conv, tableId, spSchema.ColIds)
+		// Extract common spColds.
+		colIds := GetCommonColumnIds(conv, tableId, spSchema.ColIds)
```
Comment (manitgupta) [inline sources/common/infoschema.go:124]: For my understanding, can you explain what's happening here? 
(Maybe also add a comment on why this is needed)

File: ui/src/app/components/add-new-column/add-new-column.component.html [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        </mat-form-field>
+        <br>
+        <mat-form-field appearance="outline">
+            <mat-label>Datatype</mat-label>
+            <mat-select [(ngModel)]="selectedDatatype" class="input-field" formControlName="datatype" required="true" (selectionChange)="changeValidator()">
+                <mat-option *ngFor="let datatype of datatypes" [value]="datatype">{{datatype}}</mat-option>
+            </mat-select>
+        </mat-form-field>
+        <br>
+        <div *ngIf="selectedDatatype ==='STRING' || selectedDatatype ==='VARCHAR' || selectedDatatype ==='BYTES'">
```
Comment (manitgupta) [inline ui/src/app/components/add-new-column/add-new-column.component.html:16]: Nit: Can we move this logic out to `.ts` file as a method? Long conditions inside the HTML makes reduces readability.

File: ui/src/app/components/add-new-column/add-new-column.component.html [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            <mat-form-field class="full-width" appearance="outline">
+                <mat-label>Length</mat-label>
+                <input matInput placeholder="Length" type="text" formControlName="length"/>
+            </mat-form-field>
+            <br>
+        </div>
+        <mat-form-field appearance="outline">
+            <mat-label>IsNullable</mat-label>
+            <mat-select [(ngModel)]="selectedNull" formControlName="isNullable" required="true">
+                <mat-option *ngFor="let element of isNullable" [value]="element.value">
```
Comment (manitgupta) [inline ui/src/app/components/add-new-column/add-new-column.component.html:26]: Nit: `isNullable` does not seem appropriate naming for something that iterable.

File: ui/src/app/components/add-new-column/add-new-column.component.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material/dialog';
+import { ColLength, Dialect } from 'src/app/app.constants';
+import { IColumn, ISpannerDetails } from 'src/app/model/conv';
+import { IAddColumn } from 'src/app/model/update-table';
+import { DataService } from 'src/app/services/data/data.service';
+import { TargetDetailsFormComponent } from '../target-details-form/target-details-form.component';
+interface IAddColumnProps {
+  dialect: string
+  tableId: string
+}
```
Comment (manitgupta) [inline ui/src/app/components/add-new-column/add-new-column.component.ts:12]: Do we follow a convention on where to declare interfaces?

File: ui/src/app/components/add-new-column/add-new-column.component.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  isNullable = [
+    { value: false, displayName: 'No' },
+    { value: true, displayName: 'Yes' },
+  ]
+
+  ngOnInit(): void {
+    if (this.dialect == Dialect.GoogleStandardSQLDialect) {
+      this.datatypes = ['BOOL','BYTES','DATE','FLOAT64','INT64','STRING', 'TIMESTAMP', 'NUMERIC', 'JSON']
+    } else {
+      this.datatypes = ['BOOL','BYTEA','DATE','FLOAT8','INT8','VARCHAR', 'TIMESTAMPTZ', 'NUMERIC', 'JSONB']
```
Comment (manitgupta) [inline ui/src/app/components/add-new-column/add-new-column.component.ts:52]: Should we be getting this data from an API call?

File: webv2/table/add_column.go [hunk 7]
```diff
@@ -39,3 +54,45 @@ func addColumn(tableId string, colId string, conv *internal.Conv) {
 
 	conv.SpSchema[tableId] = sp
 }
+
+func AddNewColumn(w http.ResponseWriter, r *http.Request) {
+	log.Println("request started", "method", r.Method, "path", r.URL.Path)
```
Comment (manitgupta) [inline webv2/table/add_column.go:59]: I don't think this would actually log anything, can you check?

File: webv2/table/add_column.go [hunk 8]
```diff
@@ -39,3 +54,45 @@ func addColumn(tableId string, colId string, conv *internal.Conv) {
...
+		if strings.EqualFold(c.Name, details.Name) {
+			http.Error(w, fmt.Sprintf("Multiple columns with similar name cannot exist for column : %v", details.Name), http.StatusBadRequest)
+			return
+		}
+	}
+	usedNames := internal.ComputeUsedNames(sessionState.Conv)
+	_, found := usedNames[strings.ToLower(details.Name)]
+	if found {
+		http.Error(w, fmt.Sprintf("Name '%v' is in used names, please use a different column name", details.Name), http.StatusBadRequest)
+		return
```
Comment (manitgupta) [inline webv2/table/add_column.go:85]: What is a "used name", the error message is not immediately clear I think.

**Review by manitgupta — COMMENTED** [review]
Added some minor comments

File: internal/helpers.go (hunk 1 above)
Comment (shreyakhajanchi) [reply internal/helpers.go:144]: made the changes to do it throughout the code. It was buggy before this PR. It had .lower() missing at some places

File: sources/common/infoschema.go (hunk 2 above)
Comment (shreyakhajanchi) [reply sources/common/infoschema.go]: added

File: ui/src/app/components/add-new-column/add-new-column.component.ts (hunk 5 above)
Comment (shreyakhajanchi) [reply ui/src/app/components/add-new-column/add-new-column.component.ts]: Moved to models

File: ui/src/app/components/add-new-column/add-new-column.component.ts (hunk 6 above)
Comment (shreyakhajanchi) [reply ui/src/app/components/add-new-column/add-new-column.component.ts]: moved it to app.constants

File: webv2/table/add_column.go (hunk 8 above)
Comment (shreyakhajanchi) [reply webv2/table/add_column.go]: rephrased

**Review by manitgupta — APPROVED** [review]

**Review by shreyakhajanchi — COMMENTED** [review]
Created b/285850167 to sanitise the names. Will take it up separately

**Pull request merged**